<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
<script type="text/javascript" src="../static/js/load_data_form.js"></script>

{% extends 'base.html' %}

{% block head %}
<title>How to run SPIRAL?</title>
{% endblock %}

{% block content %}
<center><h1>How to run SPIRAL?</h1></center>
<div id="loading"></div>
<div id="content">
    <h2 style="font-size:1.6vw">What will you need?</h2>
    <ol>
        <li>Gene expression data, in <strong>one</strong> of the following formats:
            <ul>
                <li>csv or txt file. In this case:
                    <ul><small>
                        <li>Gene names should appear in the first column. To be compatible with GOrilla (for the
                            enrichment
                            step), the supported formats are: gene symbol (preferred), gene and protein RefSeq,
                            Uniprot,
                            Unigene and Ensembl.
                        </li>
                        <li>Sample names should appear in the first row. If your dataset has labels (for example:
                            'day0',
                            'day1'...) the labels should appear at the beginning of each sample name, before an
                            underscore.
                            For example: 'day1_ACGTGTGA'.
                        </li>
                        <li>If your data has less than 100 cells\spots\samples, SPIRAL will run on the original
                            data. If
                            your data has more than that, SPIRAL will run in a repcell-mode.
                        </li>
                        <li>
                            Count matrix files for example:
                            <ul>
                                <li>Single cell RNA-seq data set of lymphoblastoid cells (<a target="_blank"
                                                                                             href="https://www.sciencedirect.com/science/article/pii/S1097276518308803?via%3Dihub">Zhang
                                    et al. 2019</a>) -
                                    <td><a href="{{ url_for('download_count_matrix', data_n=1) }}">count matrix</a>
                                    </td>
                                </li>
                                <li>Spatial transcriptomics data of a sagittal-posterior section of a mouse brain
                                    (<a
                                            target="_blank"
                                            href="https://www.10xgenomics.com/resources/datasets/
mouse-brain-serial-section-2-sagittal-posterior-1-standard-1-1-0">10x Genomics</a>) -
                                    <td><a href="{{ url_for('download_count_matrix', data_n=2) }}">count matrix</a>
                                    </td>
                                </li>
                                <li>Single cell RNA-seq data set of Zebrafish differentiation at 7 time points (<a
                                        target="_blank"
                                        href="https://www.science.org/doi/full/10.1126/science.aar4362">Wagner
                                    et al. 2018</a>) -
                                    <td><a href="{{ url_for('download_count_matrix', data_n=3) }}">count matrix</a>
                                    </td>
                                </li>
                            </ul>
                        </li>
                    </small></ul>
                </li>
                <br>
                <li>Feature-barcode matrix (as downloaded from the Visium website). In this case, you should have three files:
                    <ul><small>
                        <li>Counts: the file type should be mtx.gz.
                        </li>
                        <li>Features: the file type should be tsv.gz.
                        </li>
                        <li>Barcodes: the file type should be tsv.gz.
                        </li>
                    </small></ul>
                    More information regarding feature-barcode matrices can be found
                    <a target="_blank" href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/matrices">here</a>.
                </li>
            </ul>
        </li>
        <br><br>
        <li><strong>Optional:</strong> spatial coordinates file (for spatial data sets), in <strong>one</strong> of the following formats:
            <ul>
                <li>csv or txt file, designed as follows:
                    <ul><small>
                        <li>The number of rows should be the same as the number of columns in the count matrix (a row
                            for every spot\cell\sample).
                        <li>There should be two or three columns:
                            <ul>
                                <li>Two column version (the order of rows should match the order of columns in the count
                                    matrix):
                                    <ul>
                                        <li>First column: X coordinates.</li>
                                        <li>Second column: Y coordinates.</li>
                                    </ul>
                                </li>
                                <li>Three column version:
                                    <ul>
                                        <li>First column: Sample names. If the sample names fit the column
                                            names in
                                            the
                                            count
                                            matrix, we will use that. Otherwise, we will assume that the order of rows
                                            match the
                                            order of
                                            columns in the count matrix.
                                        </li>
                                        <li>Second column: X coordinates.</li>
                                        <li>Third column: Y coordinates.</li>
                                    </ul>
                                </li>

                            </ul>
                        </li>
                        <li>Spatial coordinates file for example:
                            <ul>
                                <li>Spatial transcriptomics data of a sagittal-posterior section of a mouse brain (<a
                                        target="_blank"
                                        href="https://www.10xgenomics.com/resources/datasets/
        mouse-brain-serial-section-2-sagittal-posterior-1-standard-1-1-0">10x Genomics</a>) -
                                    <td><a href="{{ url_for('download_spatial_coors', data_n=2) }}">spatial coordinates
                                        file</a>
                                    </td>
                                </li>

                            </ul>
                        </li>
                    </small></ul>
                </li>
                <br>
                <li>
                    Visium's 'tissue_positions.csv'. This file is usually located inside the 'Spatial imaging data'
                    folder, that can be downloaded for each dataset from the Visium website.
                </li>
            </ul>
        <br><br>
        <li>A local computer or a server to run SPIRAL on, with operating system Windows or Linux 64-bit. <br>
            If possible, it is best to choose a high memory
            machine. However, SPIRAL can also run on low-memory machines, assuming you use a smaller Nc
            (the number of repcells). You will be asked if you want to change this parameter during the running process.
        </li>
    </ol>

    <br><br>
    <h2 style="font-size:1.6vw">What should you do?</h2>
    In short: download spiral.zip, unzip it and run SPIRAL with "spiral.exe" (in Windows) or "spiral" (in Linux).
    You can do this manually or programmatically:
    <ul>
        <li>Manually:
            <ul>
                <li>Choose the correct spiral.zip version based on the operating system of your server and download it here:
                    <ul>
                        <li>Windows: <td><a href="{{ url_for('download_spiral_zip', sys='windows') }}">spiral.zip</a></td>
                        </li>
                        <li>Linux 64-bit: <td><a href="{{ url_for('download_spiral_zip', sys='linux_64') }}">spiral.zip</a></td>
                        </li>
                    </ul>
                </li>
                <li>Right-click spiral.zip and choose "unzip/extract here".</li>
                <li>Inside the new "spiral" folder, double-click "spiral.exe" (in Windows) or "spiral" (in Linux).</li>
                <li>Follow the instructions.</li>
            </ul>
        </li>
        <br>
        <li>Programmatically:
            <ul>
                <li>windows:
                    <ul>
                        <li>Open the cmd (locally) or connect to a remote server.</li>
                        <li>Download spiral.zip for Windows:
                            <code>curl -o spiral.zip https://spiral.technion.ac.il/download_spiral_zip_windows</code></li>
                        <li>Unzip spiral.zip: <code>tar -xf spiral.zip</code></li>
                        <li>Enter the new "spiral" folder: <code>cd spiral</code></li>
                        <li>run "spiral.exe": <code>spiral.exe</code></li>
                        <li>Follow the instructions in the cmd.</li>
                    </ul>
                </li>
                <li>Linux:
                    <ul>
                        <li>Open the terminal (locally) or connect to a remote server.</li>
                        <li>Download spiral.zip for Linux 64-bit:
                            <code>wget --content-disposition --no-check-certificate https://spiral.technion.ac.il/download_spiral_zip_linux_64</code></li>
                        <li>Unzip spiral.zip: <code>unzip spiral.zip</code></li>
                        <li>Enter the new "spiral" folder: <code>cd spiral</code></li>
                        <li>run "spiral": <code>./spiral</code><br>
                        Note: if you get a permission error, first run: <code>chmod -R 0777 ./</code></li>
                        <li>Follow the instructions in the terminal.</li>
                    </ul>
                </li>
            </ul>
        </li>
    </ul>

    <h2 style="font-size:1.6vw; color:#A981BB;">When done, you can upload the spiral_results.zip file to view the results, as explained
    <a target="_blank" href="{{ url_for('how_to_view') }}">here</a>.</h2>

</div>
{% endblock %}