{% extends 'base.html' %}

{% block head %}
<title>Home | SPIRAL</title>
{% endblock %}

{% block content %}
<center><h1>Welcome to SPIRAL</h1></center>
<br>
<p>SPIRAL is an algorithm that relies on a Gaussian statistical model to produce a comprehensive overview of significant
    processes in single cell RNA-seq, spatial transcriptomics or bulk RNA-seq. SPIRAL detects biological processes by
    identifying the subset of genes involved and the subset of cells, spots or samples. </p>

<h2 style="font-size:1.6vw">SPIRAL output for different data sets:</h2>
<ul>
    <li>Single cell RNA-seq data set of lymphoblastoid cells (<a target="_blank"
                                                                 href="https://www.sciencedirect.com/science/article/pii/S1097276518308803?via%3Dihub">Zhang
        et al. 2019</a>) - <a
            href={{ Zhang2019_link }}>SPIRAL results</a>
    </li>

    <li>Single cell RNA-seq data set of Zebrafish differentiation at 7 time points (<a target="_blank"
                                                                                       href="https://www.science.org/doi/full/10.1126/science.aar4362">Wagner
        et al. 2018</a>) - <a
            href={{ Wagner2018_link }}>SPIRAL results</a>
    </li>

    <li>Spatial transcriptomics data of a sagittal-posterior section of a mouse brain (<a target="_blank"
                                                                                          href="https://www.10xgenomics.com/resources/datasets/
mouse-brain-serial-section-2-sagittal-posterior-1-standard-1-1-0">10x Genomics</a>) - <a
            href={{ MouseSP2_link }}>SPIRAL results</a>
    </li>

    <li>Spatial transcriptomics data of a normal human prostate (<a target="_blank"
                                                                    href="https://www.10xgenomics.com/resources/datasets/normal- humanprostate-
ffpe-1-standard-1-3-0">10x Genomics</a>) - <a
            href={{ normal_prostate_link }}>SPIRAL results</a>
    </li>

    <li>Bulk RNA-seq data of human differentiation- samples were taken at days 0 and 7âˆ’21 (Mandel-Gutfreund lab,
        available at <a target="_blank"
                        href="https://github.com/hadasbi/SPIRAL.web.tool/blob/main/human_differentiation_count_matrix.csv">GitHub
            repository</a>) - <a
                href={{ yaellab_differentiation_link }}>SPIRAL results</a>
    </li>

    <li>Bulk RNA-seq data of mouse B-cells that were treated with anti-IgM mAb- 25 time points in the 6 hours post
        stimulation (<a target="_blank"
                        href="https://febs.onlinelibrary.wiley.com/doi/full/10.1002/1873-3468.13757">Chiang et al.
            2020</a>) -
        <a
                href={{ bulk_mouse_Bcells_link }}>SPIRAL results</a>
    </li>
</ul>

<br>
<br>

<div class="container">
    <form method="POST" action="/run_SPIRAL_step1">
        <div class="form-row">
            <div class="form-group">
                <center>{{ form.submit(class="btn btn-primary")}}</center>
            </div>
        </div>
    </form>
</div>

<br>
<br>

<h2 style="font-size:1.2vw">How to cite?</h2>
<p style="font-size:15px">Hadas Biran, Tamar Hashimshony, Yael Mandel-Gutfreund and Zohar Yakhini. SPIRAL: Significant
    Process
    InfeRence
    Algorithm for single cell
    RNA-sequencing and spatial transcriptomics. (2023). (in review)</p>
<div style="text-align: right;"><img src="/static/Technion_logo.jpg" width="100" height="40" alt="Technion logo"></div>
{% endblock %}