<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
<script type="text/javascript" src="../static/js/load_data_form.js"></script>

{% extends 'base.html' %}

{% block head %}
<title>Step 1 | SPIRAL</title>
{% endblock %}

{% block content %}
<center><h1>Run SPIRAL!</h1></center>
<div id="loading"></div>
<div id="content">
    <div class="formdesign">
        <h2>Step 1/5: data loading</h2>
        <form method="POST" action="" enctype="multipart/form-data" onsubmit="loading();">
            {{ form.hidden_tag() }}
            <div class="form-horizontal">
                <div class="form-group col-md-6">
                    {{ form.csrf_token() }}
                    <label for=""> {{ form.dataset_name.label }}</label>
                    {{ form.dataset_name }}
                    <p><small>This field is optional. We recommend that you name your data set, especially if you plan
                        to run
                        SPIRAL on more than one dataset. </small></p>
                </div>

                <hr>

                <div class="form-group col-md-6">
                    {{ form.csrf_token() }}
                    <label for=""> {{ form.email.label }}</label>
                    {{ form.email }}
                    <p><small>This field is optional. If filled, a link to the results page will be sent to this
                        address. </small></p>
                    {% if form.email.errors %}
                    <ul class="errors">
                        {% for error in form.email.errors %}
                        <li>{{ error }}</li>
                        {% endfor %}
                    </ul>
                    {% endif %}
                </div>

                <hr>

                <div class="form-group col-md-6">
                    {{ form.csrf_token() }}
                    <label for=""> {{ form.species.label }}</label>
                    {{ form.species }}
                    {% if form.species.errors %}
                    <ul class="errors">
                        {% for error in form.species.errors %}
                        <li>{{ error }}</li>
                        {% endfor %}
                    </ul>
                    {% endif %}
                </div>

                <hr>

                <div class="form-group col-md-6">
                    {{ form.csrf_token() }}
                    <label for=""> {{ form.samp_name.label }}</label>
                    {{ form.samp_name }}
                    {% if form.samp_name.errors %}
                    <ul class="errors">
                        {% for error in form.samp_name.errors %}
                        <li>{{ error }}</li>
                        {% endfor %}
                    </ul>
                    {% endif %}
                </div>

                <hr>
                <div id="data_types_selecting" style="margin-bottom: 10px">
                    <label for="">What is the data type?</label>
                    <div style="margin-top: 5px">
                        <input type="radio" name="data_type" value="count-matrix" id="count-matrix">
                        <label for="count-matrix">Count Matrix</label>
                    </div>
                    <div>
                        <input type="radio" name="data_type" value="visium-10x" id="visium-10x">
                        <label for="visium-10x">Visium 10X</label>
                    </div>
                </div>

                <div id="data_type_warning"><span style="color: #A52A2A">You must choose a data type in order to load your data</span>
                </div>


                <div id="count-matrix-input" class="form-group col-md-6" style="display: none">
                    {{ form.csrf_token() }}
                    <label for=""> {{ form.count_matrix.label }}</label>
                    {{ form.count_matrix }}
                    <p>
                    <ul><small>
                        <li>The file type should be either csv or txt.</li>
                        <li>Gene names should appear in the first column. To be compatible with GOrilla (for the
                            enrichment
                            step), the supported formats are: gene symbol (preferred), gene and protein RefSeq,
                            Uniprot,
                            Unigene and Ensembl.
                        </li>
                        <li>Sample names should appear in the first row. If your dataset has labels (for example:
                            'day0',
                            'day1'...) the labels should appear at the beginning of each sample name, before an
                            underscore.
                            For example: 'day1_ACGTGTGA'.
                        </li>
                        <li>If your data has less than 100 cells\spots\samples, SPIRAL will run on the original
                            data. If
                            your data has more than that, SPIRAL will run in a repcell-mode.
                        </li>
                        <li>
                            Count matrix files for example:
                            <ul>
                                <li>Single cell RNA-seq data set of lymphoblastoid cells (<a target="_blank"
                                                                                             href="https://www.sciencedirect.com/science/article/pii/S1097276518308803?via%3Dihub">Zhang
                                    et al. 2019</a>) -
                                    <td><a href="{{ url_for('download_count_matrix', data_n=1) }}">count matrix</a>
                                    </td>
                                </li>
                                <li>Spatial transcriptomics data of a sagittal-posterior section of a mouse brain
                                    (<a
                                            target="_blank"
                                            href="https://www.10xgenomics.com/resources/datasets/
mouse-brain-serial-section-2-sagittal-posterior-1-standard-1-1-0">10x Genomics</a>) -
                                    <td><a href="{{ url_for('download_count_matrix', data_n=2) }}">count matrix</a>
                                    </td>
                                </li>
                                <li>Single cell RNA-seq data set of Zebrafish differentiation at 7 time points (<a
                                        target="_blank"
                                        href="https://www.science.org/doi/full/10.1126/science.aar4362">Wagner
                                    et al. 2018</a>) -
                                    <td><a href="{{ url_for('download_count_matrix', data_n=3) }}">count matrix</a>
                                    </td>
                                </li>
                            </ul>
                        </li>
                    </small></ul>
                    </p>
                    {% if form.count_matrix.errors %}
                    <ul class="errors">
                        {% for error in form.count_matrix.errors %}
                        <li>{{ error }}</li>
                        {% endfor %}
                    </ul>
                    {% endif %}
                </div>
                <div class="form-group col-md-6" id="visium-10x-input" style="display: none">
                    {{ form.csrf_token() }}
                    <label for=""> {{ form.visium_10x_matrix.label }}</label>
                    {{ form.visium_10x_matrix }}
                    <p>
                        <small>
                            The file type should be mtx.gz.
                        </small>
                    </p>
                    {% if form.visium_10x_matrix.errors %}
                    <ul class="errors">
                        {% for error in form.visium_10x_matrix.errors %}
                        <li>{{ error }}</li>
                        {% endfor %}
                    </ul>
                    {% endif %}

                    {{ form.csrf_token() }}
                    <label for=""> {{ form.visium_10x_features.label }}</label>
                    {{ form.visium_10x_features }}
                    <p>
                        <small>
                            The file type should be tsv.gz.
                        </small>
                    </p>
                    {% if form.visium_10x_features.errors %}
                    <ul class="errors">
                        {% for error in form.visium_10x_features.errors %}
                        <li>{{ error }}</li>
                        {% endfor %}
                    </ul>
                    {% endif %}

                    {{ form.csrf_token() }}
                    <label for=""> {{ form.visium_10x_barcodes.label }}</label>
                    {{ form.visium_10x_barcodes }}
                    <p>
                        <small>
                            The file type should be tsv.gz.
                        </small>
                    </p>
                    {% if form.visium_10x_barcodes.errors %}
                    <ul class="errors">
                        {% for error in form.visium_10x_barcodes.errors %}
                        <li>{{ error }}</li>
                        {% endfor %}
                    </ul>
                    {% endif %}
                    <p>
                        <small>
                            For more information regarding Feature-Barcode Matrices see <a target="_blank"
                                                                                           href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/matrices">here.</a>
                        </small>
                    </p>
                </div>

                <hr>

                <div class="form-group col-md-6">
                    {{ form.csrf_token() }}
                    <label for=""> {{ form.labels_checkbox.label }}</label>
                    {{ form.labels_checkbox }}
                </div>

                <hr>

                <div class="form-group col-md-6">
                    {{ form.csrf_token() }}
                    <label for=""> {{ form.spatial_coors.label }}</label>
                    {{ form.spatial_coors }}
                    <p>
                    <ul><small>
                        <li>The field is optional.</li>
                        <li>The file type should be either csv or txt.</li>
                        <li>The number of rows should be the same as the number of columns in the count matrix (a row
                            for every spot\cell\sample).
                        <li>There should be two or three columns:
                            <ul>
                                <li>Two column version (the order of rows should match the order of columns in the count
                                    matrix):
                                    <ul>
                                        <li>First column: X coordinates.</li>
                                        <li>Second column: Y coordinates.</li>
                                    </ul>
                                </li>
                                <li>Three column version:
                                    <ul>
                                        <li>First column: Sample names. If the sample names fit the column
                                            names in
                                            the
                                            count
                                            matrix, we will use that. Otherwise, we will assume that the order of rows
                                            match the
                                            order of
                                            columns in the count matrix.
                                        </li>
                                        <li>Second column: X coordinates.</li>
                                        <li>Third column: Y coordinates.</li>
                                    </ul>
                                </li>

                            </ul>
                        </li>
                        <li>Spatial coordinates file for example:
                            <ul>
                                <li>Spatial transcriptomics data of a sagittal-posterior section of a mouse brain (<a
                                        target="_blank"
                                        href="https://www.10xgenomics.com/resources/datasets/
mouse-brain-serial-section-2-sagittal-posterior-1-standard-1-1-0">10x Genomics</a>) -
                                    <td><a href="{{ url_for('download_spatial_coors', data_n=2) }}">spatial coordinates
                                        file</a>
                                    </td>
                                </li>

                            </ul>
                        </li>
                    </small></ul>
                    </p>
                    {% if form.spatial_coors.errors %}
                    <ul class="errors">
                        {% for error in form.spatial_coors.errors %}
                        <li>{{ error }}</li>
                        {% endfor %}
                    </ul>
                    {% endif %}
                </div>

                <hr>

                <div class="form-group">
                    {{ form.csrf_token() }}
                    {{ form.submit(class="submitbutton")}}
                </div>


            </div>
        </form>
    </div>
</div>
{% endblock %}